nominal/core/dataset_file.py

```diff
@@ -36,8 +52,138 @@ def _from_conjure(cls, dataset_file: scout_catalog.DatasetFile) -> Self:
             uploaded_at=upload_time,
             ingested_at=ingest_time,
             ingest_status=IngestStatus._from_conjure(dataset_file.ingest_status),
+            _clients=clients,
         )
 
+    def download(
+        self,
+        output_directory: pathlib.Path,
+        force: bool = False,
+    ) -> pathlib.Path:
+        """Download the dataset file to a destination on local disk.
+
+        Args:
+            output_directory: Download file to the given directory
+            force: If true, delete any files that exist / create parent directories if nonexistant
+
+        Returns:
+            Path that the file was downloaded to
+
+        Raises:
+            FileNotFoundError: Output directory doesn't exist and force=False
+            FileExistsError: File already exists at destination
+            ValueError: Output directory exists and is not a directory
+            RuntimeError: Error downloading file
+        """
+        if output_directory.exists() and not output_directory.is_dir():
+            raise ValueError(f"Output directory is not a directory: {output_directory}")
+
+        api_uri = self._clients.catalog.get_dataset_file_uri(self._clients.auth_header, self.dataset_rid, self.id)
+        destination = output_directory / filename_from_uri(api_uri.uri)
+
+        logger.info("Downloading %s (%s) => %s", self.name, api_uri.uri, destination)
+        download_presigned_uri(api_uri.uri, destination, force=force)
+        return destination
+
+    @ttl_cache(ttl=30.0)
+    def _presigned_origin_files(self) -> Sequence[scout_catalog.OriginFileUri]:
+        return self._clients.catalog.get_origin_file_uris(self._clients.auth_header, self.dataset_rid, self.id)
+
+    def _download_origin_file(self, origin_path: str, destination: pathlib.Path, force: bool, num_retries: int) -> None:
+        last_exception = None
+        for attempt in range(num_retries):
+            origin_uris = self._presigned_origin_files()
+            origin_uri = None
+            for api_uri in origin_uris:
+                if api_uri.path == origin_path:
+                    origin_uri = api_uri.uri
+                    break
+
+            if origin_uri is None:
+                raise RuntimeError(f"No such origin path: {origin_path}")
```

Comment:
can we just return a dict or something instead of needing to have an optional that is later assigned and checked?